### ISSUE TITLE
Community releases of bitkey should be signed

### ISSUE THREAD

**Issue by matiaslb** [bitkey/bitkey#57]
It isn't clear whether upcoming releases in the community branch will be signed using the "Turnkey Linux Release Key <release@turnkeylinux.com>" key ([link](https://releases.turnkeylinux.org/turnkey-bitkey/14.2.0-jessie-amd64/) to the signatures of version 14.2.0). So opening this issue as a reminder that as a user I'd like to verify that I'm getting the same artifacts as the ones that were released by the maintainers.

Here's a quick tutorial on two ways to go about embedding the signatures in the releases page in github: https://wiki.debian.org/Creating%20signed%20GitHub%20releases

**Comment by estevaocm** [comment]
Hello, @matiaslb. That link does not hold signatures, only a digest (hash code), a manifest and a change log (which are a bit simpler to make), but I see your point. I'm getting a forbidden error from the second link you offered.

**Comment by matiaslb** [comment]
Correct, the [hash](https://releases.turnkeylinux.org/turnkey-bitkey/14.2.0-jessie-amd64/turnkey-bitkey-14.2.0-jessie-amd64.iso.hash) is signed, so one can run: `$ gpg --verify turnkey-bitkey-14.2.0-jessie-amd64.iso.hash` , to verify that the corresponding image hasn't been tampered with.

It's odd that you're having trouble accessing the debian wiki, here's the [link](https://webcache.googleusercontent.com/search?q=cache:sC3rCo0nNdgJ:https://wiki.debian.org/Creating%2520signed%2520GitHub%2520releases+&cd=1&hl=en&ct=clnk&gl=us&client=firefox-b-1-ab) Google cache version. It has detailed steps with screenshots, but it basically boils down to what you would expect:
* sign the release tag
* create a detached signature of the release binary with gpg
* upload the signature along with the release binaries

**Comment by estevaocm** [comment]
Right, sorry, seems like I forgot to open the file to check the PGP.

**Comment by estevaocm** [comment]
I do not currently have Turnkey's private key so I'll have to check with @JedMeister what we can do about this.

**Comment by estevaocm** [comment]
I could easily provide the ISO hashes, but I understand that would not be secure enough for some.

**Comment by matiaslb** [comment]
I'm not very familiar with key management, but I believe using your own key to sign the releases would be sufficient. 

I guess for bonus points, you could upload your public key to keybase and request [turnkeylinux](https://keybase.io/turnkeylinux) to follow you to establish a [web of trust](https://keybase.io/docs/server_security/following).

**Comment by quantumproducer** [comment]
+1

**Comment by JedMeister** [comment]
@matiaslb 
> I'm not very familiar with key management, but I believe using your own key to sign the releases would be sufficient.
>
> I guess for bonus points, you could upload your public key to keybase and request turnkeylinux to follow you to establish a web of trust.

Great suggestion! Yes that's a very good idea. Unfortunately I don't have access to TurnKey's private keys myself either (for security only @alonswartz has access to those keys; just the way I like it! :smile:). But I'm sure that we can work something out. I'll speak to Alon about it.

**Comment by JedMeister** [comment]
@estevaocm - in the meantime, I have just followed you on Keybase.io. It still doesn't quite "cut the mustard" (especially considering that TurnKey doesn't follow me!) but it's a start.

FYI, [buildtasks](https://github.com/turnkeylinux/buildtasks) includes a [generate-signature](https://github.com/turnkeylinux/buildtasks/blob/master/bin/generate-signature) script. The resulting hash file, then needs to be signed with your key. Having said that, the generated file will have all the info relating to TurnKey's key, so that won't really do... TBH, I'm not sure on the best way of resolving that. Obviously we'll need to have a slightly different process... I'm open to ideas and suggestions...

**Comment by estevaocm** [comment]
I'll look into signing with my Keybase key. But I'm maintaining Bitkey on a Windows machine with VMs for building and testing, and PGP tools haven't so far been very convenient in this system. Also the complexity of feature requests has been escalating somewhat, so I won't be able to release as often as I was doing in my fork that became this community version. I am preparing a new release in which I will try to upgrade Python beyond Jessie's package and Electron Cash with it; if I succeed, that may be the first signed release.

**Comment by matiaslb** [comment]
Perhaps we should consider adding PGP tools to bitkey so that you can sign the releases in cold-offline mode... 🤔 

I'm only half joking though, I wonder if there would be interest in expanding bitkey a little beyond being the "swiss army knife of handy Bitcoin tools" and to include tools that facilitate other sensitive workflows. Just a thought.

Edit: Perhaps I'll give it a shot this weekend, I've been thinking about taking better care of my private keys for quite some time. I'll submit a PR if I end up with a sufficiently streamlined workflow.

**Comment by JedMeister** [comment]
@estevaocm - we don't want to make this too much work for you! So at the end of the day, please make sure you don't burn yourself out by taking on too much!

Seeing as you are doing all of this in your own time, we'd rather you keep the project alive, even if it doesn't have all the bells and whistles that everybody wants! 

Having said that, a signed hash file would be really nice to have.

Regarding creating the hash file, as I noted, I'm sure that we could bend buildtasks (on TKLDev) to create the hash file (if you don't already have something set up).

For the signing part, again setting that up on TKLDev is one possibility. You'd need to import your GPG/PGP key pair into your TKLDev, but then you should be able to do it there relatively easily (using commandline tools). It may take a little mucking around to set it up, but AFAIK it should be quite possible. I'm happy to give you some assistance if required.

Also, another possibility is using the Keybase app (they have Windows builds - if you don't already have it installed locally). It seems like it now provides "[KBFS](https://keybase.io/docs/kbfs)" (Keybase filesystem). KBFS is essentially a virtual filesystem (i.e. a Windows "drive") which auto signs any files dropped into it. Files dropped into the "public" directory are signed and then publicly available via a https://keybase.pub/ subdirectory which includes your Keybase username. Perhaps that's worth a try?

Alternatively, the Keybase app probably supports signing local files too. You only need to sign the hash file (it contains the SHA sums of the ISO, so the ISO itself doesn't need signing). And the hash file is a pretty small text file, so hopefully that should be fairly straight forward. Please feel free to contact me directly if you want to discuss any of this further directly.

**Comment by estevaocm** [comment]
Frankly I expected better usability from Keybase. I've been recording the ISO hash in the release notes but given up signing in Windows. I'm considering setting up a PGP in my Ubuntu system at work and signing the hash afterwards, but I'm not exactly thrilled at this job. Probably reasonable to do for major releases but I'm not sure I'll be signing all incremental releases. But I'll wait to make up my mind after I set the first signed release up. As always, thanks for the support, @JedMeister!

**Comment by JedMeister** [comment]
@estevaocm - you're most welcome mate. 

As I'm sure I've already mentioned, I'm a full time Linux user, so I won't be able to help much with Windows stuff, but I'm sure I could help out with a Linux commandline path (inside your TKLDev).

Regardless, see how you go and always happy to help you out if need be. :smile:

**Comment by estevaocm** [comment]
After giving this some thought, I think the best solution for regular use would be to have the TKLDev build script, or a separate script in that environment, to sign product.iso right after it is built. Then I'll retrieve the signature file from the VM right after retrieving the ISO for the release, and post the text in the release notes. @JedMeister, if you have a sample script ready from another project, I'd appreciate it. Otherwise I'll figure it out with Google when I have time.

**Comment by estevaocm** [comment]
I have set up my own PGP key. Later I'll copy the files into TKLDev to start figuring out a script to sign the ISO.

**Comment by estevaocm** [comment]
I'm disappointed with gpg, I found no way to sign the existing hash. Apparently I have to sign the original file (computing the hash again, only with SHA1 instead of 256). Otherwise I'd be posting the ASC for the latest release right now.

**Comment by JedMeister** [comment]
Hey @estevaocm apologies I haven't responded sooner. I'm just trying to tidy some fairly important stuff up. Once I've done that, I'll spend a little time looking into this a bit closer. As I think I mentioned earlier, I'm pretty sure that the scripts in buildtasks should be easily bent to our purposes here.

Having said that, I don't work with gpg much day to day (other than emails and apt repos; but both are more-or-less automated after initial set up). The way we do things when we build in TurnKey, is that I generate the hash files and Alon signs them (I don't have access to the TurnKey key). Worst case, I can ask Alon for advice.

I suggest that if you are feeling frustrated with this, that you put it aside and wait until I look into things a little closer.
